binderhub/build.py

```diff
@@ -57,7 +55,7 @@ def get_cmd(self):
 
     def progress(self, kind, obj):
         """Put the current action item into the queue for execution."""
-        self.q.put_nowait({'kind': kind, 'payload': obj})
+        IOLoop.instance().add_callback(self.q.put, {'kind': kind, 'payload': obj})
```

Comment:
I'm trying to wrap my head around this. This looks like we're wanting to call on the main thread a call to put. However, put can block, and put_nowait is just put with block=False set. It's also safe to put things into queue from multiple threads. Why this change?

Reply:
This is not a stdlib `queue.Queue`, it's a tornado `queues.Queue`.

`progress` is called from a thread and `q.put` must be called from the main thread. The only threadsafe way to call these things is to place them on the main thread via `.add_callback()`.

Reply:
Aaah, I totally missed that change! This all makes sense now.